fraud/service.go

```diff
@@ -115,30 +147,190 @@ func (f *service) processIncoming(
-	if err != nil {
-		log.Error(err)
-	}
-	log.Warn("Shutting down services...")
-	return pubsub.ValidationAccept
+
+	return getAll(ctx, store, proofType)
 }
 
-func (f *service) Get(ctx context.Context, proofType ProofType) ([]Proof, error) {
+func (f *service) put(ctx context.Context, proofType ProofType, hash string, data []byte) error {
 	f.storesLk.Lock()
 	store, ok := f.stores[proofType]
 	if !ok {
 		store = initStore(proofType, f.ds)
 		f.stores[proofType] = store
 	}
 	f.storesLk.Unlock()
+	return put(ctx, store, hash, data)
+}
 
-	return getAll(ctx, store, proofType)
+func (f *service) handleFraudMessageRequest(stream network.Stream) {
+	msg := new(pb.FraudMessage)
+	_, err := serde.Read(stream, msg)
+	if err != nil {
+		stream.Reset() //nolint:errcheck
+		log.Warn(err)
+		return
+	}
+
+	errg, ctx := errgroup.WithContext(context.Background())
+	mu := sync.Mutex{}
+	for _, p := range msg.RequestedProofType {
+		proofType := toProof(p)
+		errg.Go(func() error {
+			proofs, err := f.Get(ctx, proofType)
+			if err != nil {
+				if errors.Is(err, datastore.ErrNotFound) {
+					return nil
+				}
+				return err
+			}
+
+			for _, proof := range proofs {
+				bin, err := proof.MarshalBinary()
+				if err != nil {
+					log.Warn(err)
+					continue
+				}
+				resp := &pb.ProofResponse{Type: int32(proofType), Value: bin}
+				mu.Lock()
+				msg.Proofs = append(msg.Proofs, resp)
+				mu.Unlock()
+			}
+			return nil
+		})
+	}
+
+	if err = errg.Wait(); err != nil {
+		stream.Reset() //nolint:errcheck
+		log.Error(err)
+		return
+	}
+	_, err = serde.Write(stream, msg)
+	if err != nil {
+		log.Error(err)
+	}
+}
+
+func (f *service) Sync(ctx context.Context) {
+	go f.syncFraudProofs(ctx)
+}
+
+func (f *service) syncFraudProofs(ctx context.Context) {
+	log.Debug("start fetching Fraud Proofs")
+	sub, err := f.host.EventBus().Subscribe(&event.EvtPeerIdentificationCompleted{})
+	if err != nil {
+		log.Error(err)
+		return
+	}
+	defer sub.Close()
+	connStatus := event.EvtPeerIdentificationCompleted{}
+	for i := 0; i < fraudRequests; i++ {
+		select {
+		case <-ctx.Done():
+			return
+		case e := <-sub.Out():
+			connStatus = e.(event.EvtPeerIdentificationCompleted)
+		}
+
+		if connStatus.Peer == f.host.ID() {
+			continue
+		}
+		go func(pid peer.ID) {
+			log.Debug("requesting proofs from peer ", pid)
+			f.topicsLk.RLock()
+			proofTypes := make([]int32, 0, len(f.topics))
+			for proofType := range f.topics {
+				proofTypes = append(proofTypes, int32(proofType))
+			}
+			f.topicsLk.RUnlock()
+			respProofs, err := requestProofs(ctx, f.host, pid, proofTypes)
+			if err != nil {
+				log.Error(err)
+				return
+			}
+			if len(respProofs) == 0 {
+				log.Debug("proof was not found")
+				return
+			}
+			log.Debug("got fraud proofs from peer: ", connStatus.Peer)
+			wg := sync.WaitGroup{}
+			wg.Add(len(respProofs))
+			for _, data := range respProofs {
+				go func(data *pb.ProofResponse) {
```

Comment:
I also do not understand the value of making this async

Reply:
`h, err := f.getter(ctx, proof.Height())` is a blocking operation. So, we will be blocked until Syncer will provide needed block to verify Fraud Proof